content/en/blog/_posts/2022-08-15-immutability-with-cel.md

```diff
@@ -0,0 +1,662 @@
+```
+```console
+customresourcedefinition.apiextensions.k8s.io/mapappendonlykeys.stable.example.com created
+```
+
+Creating an initial object with one key within `values` should be permitted:
+```shell
+kubectl apply -f - <<EOF
+---
+apiVersion: stable.example.com/v1
+kind: MapAppendOnlyKeys
+metadata:
+  name: testmap
+values:
+    key1: value1
+EOF
+```
+```console
+mapappendonlykeys.stable.example.com/testmap created
+```
+
+Adding new keys to the map should also be permitted:
+```shell
+kubectl apply -f - <<EOF
+---
+apiVersion: stable.example.com/v1
+kind: MapAppendOnlyKeys
+metadata:
+  name: testmap
+values:
+    key1: value1
+    key2: value2
+EOF
+```
+```console
+mapappendonlykeys.stable.example.com/testmap configured
+```
+
+But if a key is removed, the error messagr from the validation rule should be
+returned:
+```shell
+kubectl apply -f - <<EOF
+---
+apiVersion: stable.example.com/v1
+kind: MapAppendOnlyKeys
+metadata:
+  name: testmap
+values:
+    key1: value1
+EOF
+```
+```console
+The MapAppendOnlyKeys "testmap" is invalid: values: Invalid value: "object": Keys may not be removed and their values must stay the same
+```
+
+If the entire field is removed, the other validation rule is triggered and the
+operation is prevented. Note that the error message for the validation rule is
+shown to the user.
+```shell
+kubectl apply -f - <<EOF
+---
+apiVersion: stable.example.com/v1
+kind: MapAppendOnlyKeys
+metadata:
+  name: testmap
+EOF
+```
+```console
+The MapAppendOnlyKeys "testmap" is invalid: <nil>: Invalid value: "object": Value is required once set
+```
+
+### Generated schema
+
+```yaml
+openAPIV3Schema:
+  description: A map which does not allow keys to be removed or their values
+    changed once set. New keys may be added, however.
+  properties:
+    values:
+      additionalProperties:
+        type: string
+      maxProperties: 10
+      type: object
+      x-kubernetes-validations:
+      - message: Keys may not be removed and their values must stay the same
+        rule: oldSelf.all(key, key in self && self[key] == oldSelf[key])
+  type: object
+  x-kubernetes-validations:
+  - message: Value is required once set
+    rule: '!has(oldSelf.values) || has(self.values)'
+```
+
+# Going further
+The above examples showed how CEL rules can be added to kubebuilder types.
+The same rules can be added directly to OpenAPI if writing a manifest for a CRD by hand.
+
+For native types, the same behavior can be achieved using kube-openapi’s marker
+[`+validations`](https://github.com/kubernetes/kube-openapi/blob/923526ac052c59656d41710b45bbcb03748aa9d6/pkg/generators/extension.go#L69).
+
+Usage of CEL within Kubernetes Validation Rules is so much more powerful than
+what has been shown in this article. For more information please check out
+[validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
+in the Kubernetes documentation.
```

Comment:
We could also link to the blog that https://github.com/kubernetes/website/pull/35459 will publish, _and_ we could amend that article to link to this one (would need us to delay merging this PR, so that we don't link to a not-yet-published article).

Reply:
Link to a prospective URL added. Agreed hold until link is confirmed

Reply:
added a link to https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/